Tech preview Kong IDP Manual (#1626)
Kong/developer.konghq.com: app/konnect-platform/kong-identity.md

{% mermaid %}
sequenceDiagram
    autonumber
    participant client as Client Application
    participant idsvc as Kong Identity Service
    participant gateway as Kong Gateway
    participant api as Customer API
    client->>idsvc: request access token<br>(Client ID, Client Secret, requested scope)
    idsvc->>idsvc: validate credentials
    idsvc-->>client: access token, granted scope, expiry time
    client->>gateway: /get_protected_resource<br>(access token, granted scope)
    gateway->>idsvc: validate access token
    idsvc-->>gateway: validate access token
    gateway->>api: get_resource
    api-->>gateway: resource
    gateway-->>client: resource
{% endmermaid %}
<!--vale on-->

## Kong Identity authorization code flow

In the authorization code flow:
1. Optional: The client application displays the user consent page and authenticates the user (this part is handled outside {{site.base_gateway}}). When the user clicks **Authorize**, the client app calls the `/authorize` endpoint created by attaching the OAuth2 plugin to a service.
   {:.info}
   > If an app requires user authentication, the authorization step must happen outside of {{site.konnect_short_name}}.

> 3. The request includes the client ID, secret, and scopes the user consented to.

Reviewer: ```suggestion
3. The client makes a request that includes the client ID, secret, and scopes the user consented to.
```
Making this an action rather than just info.

4. The authorization server ({{site.base_gateway}} with OAuth2 plugin) validates the client credentials and returns an authorization code.
5. The client exchanges this code at the `/oauth/token` endpoint for access tokens.

> 6. The access token is used to call protected APIs.

Reviewer: ```suggestion
6. The client uses the access token to call protected APIs.
```

<!--
For Consumer Group-scoped plugins:
- Create a consumer per client in the respective control plane.
- No need to migrate the client credential to a consumer credential.
- The OIDC plugin maps clients to consumers using claims.
- Create the required consumer groups and apply the plugin at the consumer group scope.
- Add each consumer to the appropriate consumer group in the control plane.

> -->

Reviewer 2: This one is for Consumer groups, which will be coming in M2 in August. I'll move it to the GH ticket instead and delete it from here.

## Dynamic claim templates

> You can configure dynamic custom claims with dynamic claim templating to generate claims during runtime. These JWT claim values can be rendered as multiple types: 
> * Strings 
> * Integers
> * Floats
> * Booleans
> * JSON object or arrays

Reviewer: ```suggestion
You can configure dynamic claim templates to generate custom claims during runtime. These JWT claim values can be rendered as any of the following types:
* Strings 
* Integers
* Floats
* Booleans
* JSON object or arrays
```
How's this rephrase of the first sentence? The previous was a bit overwhelming with repetition.

> The type is inferred from the value. Moreover, JWT claims values can be templated with contextual data and functions. Dynamic values must use `${}` as templating boundaries.

Reviewer: ```suggestion
The type is inferred from the value. 

JWT claim values can also be templated with contextual data and functions. Dynamic values must use `${}` as templating boundaries.
```

> Claims support templating via the context passed to the client during the authentication. The context is represented by the following format:

Reviewer: ```suggestion
Claims support templating via the context passed to the client during the authentication, in the following format:
```

---
Kong Identity enables customers to use Konnect to generate, authenticate and authorize API access. Kong Identity implements the OAuth2.0 standard with OpenID Connect for authentication and authorization. Kong Identity can be used with the following Kong plugins today:
  1. OpenID Connect plugin
  2. OAuth2.0 Introspection plugin
  3. Upstream OAuth plugin

## Kong Identity Flow
<!--vale off-->
{% mermaid %}
sequenceDiagram
    autonumber
    participant client as Client Application
    participant idsvc as Kong Identity Service
    participant gateway as Kong Gateway
    participant api as Customer API
    client->>idsvc: request access token<br>(Client ID, Client Secret, requested scope)
    idsvc->>idsvc: validate credentials
    idsvc-->>client: access token, granted scope, expiry time
    client->>gateway: /get_protected_resource<br>(access token, granted scope)
    gateway->>idsvc: validate access token
    idsvc-->>gateway: validate access token
    gateway->>api: get_resource
    api-->>gateway: resource
    gateway-->>client: resource
{% endmermaid %}
<!--vale on-->

<!--vale off-->
{% konnect_api_request %}
url: /v1/auth-servers
status_code: 201
method: POST
headers:

>   - 'Authorization: Bearer $KONNECT_TOKEN'

Reviewer: ```suggestion
```
